### PR TITLE
feat(3id-did-resolver): did metadata resolution

### PR DESCRIPTION
**Pull request by oed**
This PR adds support for DIDDocumentMetadata by doing a multiquery to get the state of the 3ID DID document at two points in time. This is supported using both the `versionId` and `versionTime` parameters.

This should bring this resolver up to date with the [3ID DID Method Specification](https://github.com/ceramicnetwork/CIP/blob/main/CIPs/CIP-79/CIP-79.md)

I included the `generateVectorDocuments.js` script which was used to generate the test vectors (deps not included). I also included the commit logs for the two main documents that we are testing with.

The tests mock the ceramic daemon responses that are expected for specific http api calls.

### CODE REVIEW

**Comment by oed** [comment]
@stbrody looks like the same circleci test issue you saw earlier.

**Comment by stbrody** [comment]
> @stbrody looks like the same circleci test issue you saw earlier.

PR for the fix: https://github.com/ceramicnetwork/js-ceramic/pull/1136

File: packages/3id-did-resolver/src/index.ts [hunk 1]
```diff
@@ -53,80 +61,128 @@ export function wrapDocument(content: any, did: string): DIDDocument {
   }, startDoc)
 }
 
-const DID_LD_JSON = 'application/did+ld+json'
-const DID_JSON = 'application/did+json'
+function extractMetadata(resolvedState: DocState, fullState: DocState): DIDDocumentMetadata {
```
Comment (stbrody) [inline packages/3id-did-resolver/src/index.ts:64]: Took me a long time to figure out the difference between "resolvedState" and "fullState".  Can we call them "requestedVersionState" and "currentVersionState"?

And also add a doccomment here?

**Review by stbrody — APPROVED** [review]
code changes lgtm, a couple of small suggestions to make things clearer to readers of the code

File: packages/3id-did-resolver/generateVectorDocuments.js [hunk 2]
```diff
@@ -0,0 +1,178 @@
+const Ceramic = require('@ceramicnetwork/http-client').default
```
[the remark this answers is not in the document]
Comment (oed) [reply packages/3id-did-resolver/generateVectorDocuments.js:1]: yep!

File: packages/3id-did-resolver/src/index.ts (hunk 1 above)
Comment (oed) [reply packages/3id-did-resolver/src/index.ts]: yep!

File: packages/3id-did-resolver/src/index.ts [hunk 3]
```diff
@@ -27,6 +27,13 @@ const isLegacyDid = (didId: string): boolean => {
   }
 }
 
+/**
+ * Wraps the content from the Ceramic 3ID docuemnt tile and formats it
```
Comment (stbrody) [inline packages/3id-did-resolver/src/index.ts:31]:
```suggestion
 * Wraps the content from the Ceramic 3ID document tile and formats it
```

File: packages/3id-did-resolver/src/index.ts [hunk 4]
```diff
@@ -61,15 +68,23 @@ export function wrapDocument(content: any, did: string): DIDDocument | null {
   }, startDoc)
 }
 
-function extractMetadata(resolvedState: DocState, fullState: DocState): DIDDocumentMetadata {
+/**
+ * Extracts the DIDDocumentMetadata for the 3ID that we have resolved.
+ * Requires the latest version of the 3ID ceramic docuemnt state as
```
Comment (stbrody) [inline packages/3id-did-resolver/src/index.ts:73]:
```suggestion
 * Requires the latest version of the 3ID ceramic document state as
```

File: packages/3id-did-resolver/src/index.ts (hunk 3 above)
Comment (oed) [reply packages/3id-did-resolver/src/index.ts]: lol, ty!

File: packages/3id-did-resolver/src/index.ts (hunk 4 above)
Comment (oed) [reply packages/3id-did-resolver/src/index.ts]: Docuemnt is a hard word to spell 😅

File: packages/3id-did-resolver/generateVectorDocuments.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (stbrody) [reply packages/3id-did-resolver/generateVectorDocuments.js:11]: Looks great, thanks!

**Pull request merged**